server/src/main/java/org/opensearch/action/admin/cluster/shards/TransportCatShardsAction.java

```diff
@@ -0,0 +1,104 @@
+import org.opensearch.tasks.Task;
+import org.opensearch.transport.TransportService;
+
+/**
+ * Perform cat shards action
+ *
+ * @opensearch.internal
+ */
+public class TransportCatShardsAction extends HandledTransportAction<CatShardsRequest, CatShardsResponse> {
+
+    private final TransportService transportService;
+    private final ClusterService clusterService;
+    private final NodeClient client;
+
+    @Inject
+    public TransportCatShardsAction(
+        NodeClient client,
+        TransportService transportService,
+        ActionFilters actionFilters,
+        ClusterService clusterService
+    ) {
+        super(CatShardsAction.NAME, transportService, actionFilters, CatShardsRequest::new);
+        this.clusterService = clusterService;
+        this.transportService = transportService;
+        this.client = client;
+    }
+
+    @Override
+    public void doExecute(Task task, CatShardsRequest shardsRequest, ActionListener<CatShardsResponse> listener) {
+        final ClusterStateRequest clusterStateRequest = new ClusterStateRequest();
+        clusterStateRequest.local(shardsRequest.getLocal());
+        clusterStateRequest.clusterManagerNodeTimeout(shardsRequest.getClusterManagerNodeTimeout());
+        clusterStateRequest.clear().nodes(true).routingTable(true).indices(shardsRequest.getIndices());
+
+        if (task != null) {
+            clusterStateRequest.setParentTask(client.getLocalNodeId(), task.getId());
+            TimeoutTaskCancellationUtility.wrapWithCancellationListener(
+                client,
+                (CancellableTask) task,
+                ((CancellableTask) task).getCancellationTimeout(),
+                ActionListener.wrap(r -> {}, e -> {})
+            );
+        }
+        CatShardsResponse catShardsResponse = new CatShardsResponse();
+        try {
+            client.admin().cluster().state(clusterStateRequest, new ActionListener<ClusterStateResponse>() {
+                @Override
+                public void onResponse(ClusterStateResponse clusterStateResponse) {
+                    catShardsResponse.setClusterStateResponse(clusterStateResponse);
+                    IndicesStatsRequest indicesStatsRequest = new IndicesStatsRequest();
+                    indicesStatsRequest.all();
+                    indicesStatsRequest.indices(shardsRequest.getIndices());
+                    indicesStatsRequest.setParentTask(client.getLocalNodeId(), task.getId());
+                    try {
+                        client.admin().indices().stats(indicesStatsRequest, new ActionListener<IndicesStatsResponse>() {
+                            @Override
+                            public void onResponse(IndicesStatsResponse indicesStatsResponse) {
+                                catShardsResponse.setIndicesStatsResponse(indicesStatsResponse);
+                                listener.onResponse(catShardsResponse);
+                            }
+
+                            @Override
```

Comment:
We can use [ActionListener#Wrap](https://github.com/rajiv-kv/OpenSearch/blob/de2b6b7a6b7cafea204f1d9b3ff0830780bafc08/server/src/main/java/org/opensearch/action/ActionRunnable.java#L85) to avoid repeating the onFailure block

Reply:
There is no timeouthandler, it's new transport call. Can you help what's the ask here?

Reply:
I see the cancel check is now moved into the method `wrapWithCancellationCheck`